### DISCUSSION TITLE
LMMS version for Android: Is it possible?

### DISCUSSION THREAD

**Discussion by Ale-XmQ** [LMMS/lmms#6658] [General]
Dear LMMS developers,

I love using LMMS on my computer to create electronic music. However, I often find myself needing to produce music while on the go, and I would like to be able to use LMMS on my Android phone.

I understand that creating an Android version of LMMS may take a lot of time and resources, but I would like to ask you to consider developing an Android version of LMMS in the future. I believe that many users would be interested in having access to a mobile version of LMMS, and I am sure that an Android version of LMMS would be well received by the user community.

I appreciate your work in creating LMMS and look forward to any updates you can provide on the possibility of an Android version of LMMS in the future.

Thank you and best regards.

**Comment by Veratil** [comment]
Hi Ale-XmQ,

We're actually in the process of reworking how LMMS is structured. This restructure is aimed at allowing multiple versions of LMMS to be built over a solid core foundation (we'll have an LMMS core which can have multiple frontends developed for, mobile being one of those frontends). A mobile frontend is actually one of the targets we've discussed.

So yes! It will be possible, although at the moment it's not. Whether or not we create an official mobile frontend is up to community support really, but getting there is going to take a bit of time as we tear apart the current LMMS and mold it into something better. 😄

**Reply by jsmtux** [reply]
Hi Veratil,

I am very curious about your current plans on the restructuring efforts. Is there anywhere where I could have a look at your current plans and progress? This kind of refactoring is something I really enjoy and I've done professionally, also on a Qt application.

As a hobby, I've been hacking away a proof of concept on separating the ui into a different compile unit, where core would only be able to interact with it through a framework agnostic interface. I have moved things like MainWindow to be private to the gui library and things like that at my fork of lmms here https://github.com/jsmtux/lmms/tree/separate_ui

This is something I will keep on working on with the expectation of eventually removing the need of QWidgets in order to compile the core library. Also only using pure interfaces from plugins, to simplify compilation. However, if there is already some ongoing effort along those lines, I would be interested in helping there.

**Reply by Veratil** [reply]
The current plans are just rework core things and removing Qt dependency from it. Progress is the master branch.

There isn't anything set in stone for this yet, but the idea is essentially what you're currently working on. We're still somewhat undecided on the best way to do it will be, but your branch is definitely a good start. :)

**Reply by jsmtux** [reply]
Brilliant! Will continue on it then.
At the moment I am focusing on removing QWidgets, rather than the whole Qt, since a good alternative that works well on mobile is QML, and would be less effort.
I'll keep you posted 🙂

**Reply by jsmtux** [reply]
So, I've been working on and off in my [separate_ui branch](https://github.com/jsmtux/lmms/tree/separate_ui) and I've reached a point where it is possible to build a new target lmms_headless that has no dependency on the QWidgets library and shares all the core code with the existing lmms target.

It is still a bit pointless, since none of the plugins build with either target, but I think the current state proves this could be a valid approach to separating core and GUI. I'll keep slowly working on this until I can also build and use the plugins with either target. My hope is that once that's done, it would be possible to start working on an alternative QML based ui, with modern looks and supporting mobile devices.

If you, @Veratil, or any other maintanier of the project could have a look at the changes I'm making and give me your opinion that would be greatly appreciated.

Cheers and keep up the good work!

**Comment by Ale-XmQ** [comment]
Thank you very much for your response and for sharing such interesting information about the evolution of LMMS!

I'm glad to hear that a mobile frontend is one of the goals you've discussed and that the current restructuring will allow multiple versions of LMMS to be built on a solid foundation. I understand that it will take time, but I'm excited to think about the possibility of having an official mobile version of LMMS in the future.

I appreciate your time and work on LMMS development, and look forward to any future updates on the development of a mobile frontend.

Thanks again and best regards!

**Comment by TheGamer1002** [comment]
Bump. Any updates on this?

**Reply by Rossmaxx** [reply]
@jsmtux is working on it and got a working prototype but it's far from being merged to master due to codebase changes. We will look into it more seriously after 1.3.

**Comment by CaBachhav** [comment]
There are 2 workaround for this. 1. Use a Linux smartphone and install LMMS on it.
2. Some linux distros can be installed on the android smartphones. Then you can install LMMS in it.
